python/pyiceberg/expressions/visitors.py

```diff
@@ -842,3 +843,41 @@ def visit_bound_predicate(self, predicate: BoundPredicate[L]) -> Set[int]:
 
 def extract_field_ids(expr: BooleanExpression) -> Set[int]:
     return visit(expr, _ExpressionFieldIDs())
+
+
+class _RewriteToDNF(BooleanExpressionVisitor[Tuple[BooleanExpression, ...]]):
+    def visit_true(self) -> Tuple[BooleanExpression, ...]:
+        return (AlwaysTrue(),)
+
+    def visit_false(self) -> Tuple[BooleanExpression, ...]:
+        return (AlwaysFalse(),)
+
+    def visit_not(self, child_result: Tuple[BooleanExpression, ...]) -> Tuple[BooleanExpression, ...]:
+        return tuple(~result for result in child_result)
```

Comment:
I think that this would violate the guarantee used by the visitor, that the child result is a tuple of expressions that contain no `Or` nodes, but may contain `And` nodes. If there is an `And(left, right)` then it is negated as `Or(Not(left), Not(right))`, which will violate the guarantee when returned.

Luckily, this won't happen because the `Not` nodes are removed when using `rewrite_to_dnf`. But I think that this method should raise an error when it is used improperly (direct use of `_RewriteToDNF`) instead of returning something possibly incorrect.

Reply:
Ah, thanks! Raising an error is indeed the best thing to do! 👍🏻